### DISCUSSION TITLE
Always NULL as reply->str

### DISCUSSION THREAD

**Discussion by martin-kufner** [redis/hiredis#1278] [Q&A]
Hey guys,

I hope that I am not stupid...
I try to access redis via hiredis.
I can actually write to redis, and get the value via redis-cli.

However trying to read reply->str ALWAYS return NULL, even on the successful SET
Same when I try to do PING, SET, SELECT, GET etc. Always NULL.

What do I do wrong?

I've installed hiredis 1.2.0 with homebrew on OSX Sonoma 14.6.1 on M2
I also compiled 1.2.1 from github.

Thanks a lot, Martin

```
static redisContext *redis_conn = NULL;  // Global Redis connection
extern char *redis_host;
extern int redis_port;
extern int redis_db;
...
redis_conn = redisConnectWithTimeout(redis_host, redis_port, timeout);
....
void access_redis() {
    redisReply *reply;
    char * redis_key = "show_id:62fda075-4c03-4b88-9894-8ba7498051b1,proceed";
    reply = redisCommand(redis, "SET %s %s", "A", redis_key);
    if (reply->type == REDIS_REPLY_STATUS && reply->str != NULL) {
        log_message(LOG_INFO_LEVEL, "[timer] Redis SET status: %s", reply->str);  // Should print "OK"
    }
   // Query Redis for the key
   reply = redisCommand(redis, "GET %s", "A");
   log_message(LOG_INFO_LEVEL, "[timer] Redis reply type: %d", reply->type);
   if (reply->str != NULL) {
       log_message(LOG_INFO_LEVEL, "[timer] Redis reply string: %s (length: %lu)", reply->str, strlen(reply->str));
   } else {
       log_message(LOG_ERROR_LEVEL, "[timer] Redis reply string is NULL.");
   }
```

**Comment by michael-grunder** [comment]
Try a simpler program first that doesn't have any other moving parts.

```c
#include <hiredis/hiredis.h>
#include <stdio.h>
#include <assert.h>

#define REDIS_HOST "127.0.0.1"
#define REDIS_PORT 6379


int main(void) {
    redisContext *c;
    redisReply *r;

    c = redisConnect(REDIS_HOST, REDIS_PORT);
    assert(c != NULL && c->err == 0);

    char *redis_key = "show_id:62fda075-4c03-4b88-9894-8ba7498051b1,proceed";

    r = redisCommand(c, "SET %s %s", "A", redis_key);
    assert(r != NULL && r->type == REDIS_REPLY_STATUS);
    printf("reply->str %s\n", r->str ? r->str : "(null)");
    freeReplyObject(r);

    r = redisCommand(c, "GET %s", "A");
    assert(r != NULL && r->type == REDIS_REPLY_STRING);
    printf("reply->str %s\n", r->str ? r->str : "(null)");
    freeReplyObject(r);

    redisFree(c);
}
```

```
cc -Wall test.c -o test -lhiredis
./test
```

Since you're getting non-null `redisReply` from `redisCommand` I wouldn't expect the context to show an error but you can always check it (`ctx->err`, and `ctx->errstr`).

You can also try monitoring the Redis instance when you run your program to see if you see the commands executing.

Another possibility is an ABI mismatch. This can happen if the hiredis headers are of a different version than the compiled shared object. It would cause weird random errors.

**Reply by martin-kufner** [reply]
Thanks, the reduced version works.

But what puzzles me, the commands are executing interact with redis with my moving parts. I can select a database, I can store keys find them in redis-cli.

I actually removed the whole hiredis all libs and includes, reinstalled it both brew and from github.

I now deinstalled the brew version, use the github version and proceed with printf-debugging - actually its not that large code base. and ->str is touched in just six positions.

Do you have a clue what I can look for? Some of those memcpy assignments, len mismatch perhaps ? 

I actually did a lot in C but unfortunately a while ago, +20 yrs ... 
Anyhow, will do a try tomorrow.

Good night,

**Comment by martin-kufner** [comment]
Found it ... Thank you, @michael-grunder 

I use Ruby C and I use the hiredis gem, which comes with with the hiredis c suite.
The ruby gem included a different version of hiredis, which conflicted with the library installed by brew.

So thanks for this hint!
> Another possibility is an ABI mismatch. This can happen if the hiredis headers are of a different version than the compiled shared object. It would cause weird random errors.

How to solve this problem
uninstall hiredis  `brew uninstall hiredis` remove the includes and lib from the hiredis source. 

In the Ruby C module adapted the extconf.rb (sorry just hard coded for now)
```
...  
  hiredis_path = "/Users/mk/.rvm/gems/ruby-3.2.4/gems/hiredis-0.6.3/vendor/hiredis/"
  $INCFLAGS << " -I#{hiredis_path}"
  $LDFLAGS  << " -L#{hiredis_path}"
  dir_config('hiredis', hiredis_path, hiredis_path)
  have_header('hiredis.h')
...
```
hires_path can be determined via ... 
`gem which hiredis` =>  /Users/mk/.rvm/gems/ruby-3.2.4@qb3/gems/hiredis-0.6.3/lib/hiredis.rb
or 
`bundle show hiredis` => /Users/mk/.rvm/gems/ruby-3.2.4@qb3/gems/hiredis-0.6.3
